File: Native/Test/Fabric/CORE-162.js
FABRIC = require('Fabric').createClient();

var ShaderSource = function( code, type ) {
	this.code = (code!=undefined) ? code : "";
	this.type = (type!=undefined) ? type : 0;
};
	
FABRIC.RegisteredTypesManager.registerType( 
	'ShaderSource', 
	{
		members: { 
			code:'String', type:'Integer' 
		},
		constructor: ShaderSource
	} 
);

var klCode = "";
klCode += "operator bar( io ShaderSource shaders[] ) {\n";
klCode += "\tfor( var Size i = 0; i < shaders.size(); i++ ) {\n";
klCode += "\t\treport( \"Type: \" + shaders[i].type );\n";
klCode += "\t\treport( \"Code: \" + shaders[i].code );\n";
klCode += "\t}\n";
klCode += "}\n";

console.log( klCode );

var sources = [];
sources.push( new ShaderSource( "Zero", 0 ) );
sources.push( new ShaderSource( "One", 1 ) );

var node = FABRIC.DependencyGraph.createNode("foo");
node.addMember( 'shaderSources', 'ShaderSource[]' );
node.setData( 'shaderSources', sources );

var op = FABRIC.DependencyGraph.createOperator("bar");
op.setEntryFunctionName("bar");
op.setSourceCode(klCode);

var binding = FABRIC.DependencyGraph.createBinding();
binding.setOperator( op );
binding.setParameterLayout( ['self.shaderSources' ] );

node.bindings.append( binding );
var errors = node.getErrors();
if ( errors.length > 0 ) {
  for ( var i in errors )
    console.log( errors[i] );
}
else {
  node.evaluate();
}

FABRIC.flush();

File: Native/Test/Python/CORE-162.py
import fabric
fabricClient = fabric.createClient()

class ShaderSource():
  def __init__( self, code = None, type = None ):
    self.code = '' if code is None else code
    self.type = 0 if type is None else type
    
fabricClient.RegisteredTypesManager.registerType( 
	'ShaderSource', 
	{
		'members': [ 
			{ 'code':'String' }, { 'type':'Integer' }
		],
		'constructor': ShaderSource
	} 
)

klCode = ""
klCode += "operator bar( io ShaderSource shaders[] ) {\n"
klCode += "\tfor( Size i = 0; i < shaders.size(); i++ ) {\n"
klCode += "\t\treport( \"Type: \" + shaders[i].type );\n"
klCode += "\t\treport( \"Code: \" + shaders[i].code );\n"
klCode += "\t}\n"
klCode += "}\n"

print( klCode )

sources = []
sources.append( ShaderSource( "Zero", 0 ) )
sources.append( ShaderSource( "One", 1 ) )

node = fabricClient.DependencyGraph.createNode("foo")
node.addMember( 'shaderSources', 'ShaderSource[]' )
node.setData( 'shaderSources', sources )

op = fabricClient.DependencyGraph.createOperator("bar")
op.setEntryFunctionName("bar")
op.setSourceCode(klCode)

binding = fabricClient.DependencyGraph.createBinding()
binding.setOperator( op )
binding.setParameterLayout( ['self.shaderSources' ] )

node.bindings.append( binding )
errors = node.getErrors()
if len( errors ) > 0:
  for i in errors:
    print( errors[i] )
else:
  node.evaluate()

fabricClient.close()
